drivers/usb/device/usb_dc_usbip.c

```diff
@@ -0,0 +1,922 @@
+			return -EIO;
+		}
+	} else {
+		if (ep_idx >= USBIP_IN_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.in_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_IN;
+		LOG_DBG("DATA IN event ep 0x%02x %u", ep, ep_ctrl->buf_len);
+
+		/* Send queued data */
+		if (!usbip_send_common(ep, ep_ctrl->buf_len)) {
+			return -EIO;
+		}
+
+		if (usbip_send(ep, ep_ctrl->buf, ep_ctrl->buf_len) !=
+		    ep_ctrl->buf_len) {
+			return -EIO;
+		}
+
+		LOG_HEXDUMP_DBG(ep_ctrl->buf, ep_ctrl->buf_len, ">");
+
+		/*
+		 * Call the callback only if data in usb_dc_ep_write()
+		 * is actually written to the intermediate buffer and sent.
+		 */
+		if (ep_ctrl->buf_len != 0) {
+			ep_ctrl->cb(ep, USB_DC_EP_DATA_IN);
+			usbip_ctrl.in_ep_ctrl[ep_idx].buf_len = 0;
+		}
+	}
+
+	return 0;
+}
+
+static int usbip_recv(uint8_t *buf, size_t len)
+{
+	return recv(usbip_ctrl.connfd, buf, len, MSG_WAITALL);
+}
+
+static int usbip_send(uint8_t ep, const uint8_t *data, size_t len)
+{
+	return send(usbip_ctrl.connfd, data, len, 0);
+}
+
+static bool usbip_send_common(uint8_t ep, uint32_t data_len)
+{
+	struct usbip_return rsp;
+	uint32_t ep_dir = USB_EP_DIR_IS_IN(ep) ? USBIP_DIR_IN : USBIP_DIR_OUT;
+	uint32_t ep_idx = USB_EP_GET_IDX(ep);
+	int rc;
+
+	rsp.hdr.command = htonl(USBIP_RET_SUBMIT);
+	rsp.hdr.seqnum = htonl(usbip_ctrl.connfd);
+	rsp.hdr.devid = htonl(0);
+	rsp.hdr.direction = htonl(ep_dir);
+	rsp.hdr.ep = htonl(ep_idx);
+
+	memset(&rsp.submit, 0, sizeof(rsp.submit));
+	rsp.submit.actual_length = htonl(data_len);
+
+	rc = send(usbip_ctrl.connfd, &rsp, sizeof(rsp), 0);
+	if (rc != sizeof(rsp)) {
+		LOG_ERR("Failed to send response header, errno %d", errno);
+		return false;
+	}
+
+	usbip_ctrl.cmd_num_out++;
+
+	return true;
+}
+
+static void usbip_init_devlist_data(void)
+{
+	struct usb_device_descriptor *dev_desc;
+	struct usb_cfg_descriptor *cfg_desc;
+
+	dev_desc = (void *)__usb_descriptor_start;
+	cfg_desc = (void *)((uint8_t *)dev_desc +
+			    sizeof(struct usb_device_descriptor));
+
+	devlist_data.idVendor = htons(dev_desc->idVendor);
+	devlist_data.idProduct = htons(dev_desc->idProduct);
+	devlist_data.bcdDevice = htons(dev_desc->bcdDevice);
+	devlist_data.bDeviceClass = dev_desc->bDeviceClass;
+	devlist_data.bDeviceSubClass = dev_desc->bDeviceSubClass;
+	devlist_data.bDeviceProtocol = dev_desc->bDeviceProtocol;
+
+	devlist_data.bConfigurationValue = cfg_desc->bConfigurationValue;
+	devlist_data.bNumConfigurations = dev_desc->bNumConfigurations;
+	devlist_data.bNumInterfaces = cfg_desc->bNumInterfaces;
+}
+
+static int usbip_handle_devlist(int connfd)
+{
+	struct usb_desc_header *desc_hdr = __usb_descriptor_start;
+	struct usb_if_descriptor *if_desc = NULL;
+	struct usbip_devlist_iface_data iface;
+	struct usbip_devlist_header rep_hdr = {
+		.version = htons(USBIP_VERSION),
+		.code = htons(USBIP_OP_REP_DEVLIST),
+		.status = 0,
+		.ndev = htonl(1),
+	};
+	int rc;
+
+	LOG_DBG("Handle OP_REQ_DEVLIST");
+
+	rc = send(connfd, &rep_hdr, sizeof(rep_hdr), 0);
```

Comment:
Same here, looping over `send()` is recommended (probably in some helper function to re-use the code across all places). Plus, sending less bytes than requested would to set the `errno` so returning it in any case is not right (errno would only be set when  `-1`.is returned). Same story in other places below.